url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/render/Rasterizer2D.java

```diff
@@ -0,0 +1,458 @@
+//===========================================================================
+//=-------------------------------------------------------------------------=
+//= References:                                                             =
+//= [BRES1965] Bresenham, J.E. "Algorithm for computer control of a digital =
+//=            plotter" IBM Syst. J. 4, 1 (1965), 25-30.                    =
+//= [FOLE1992] Foley, vanDam, Feiner, Hughes. "Computer Graphics, princi-   =
+//=            ples and practice" - second edition, Addison Wesley, 1992.   =
+//===========================================================================
+
+package vsdk.toolkit.render;
+
+import vsdk.toolkit.common.VSDK;
+import vsdk.toolkit.common.ColorRgb;
+import vsdk.toolkit.common.ArrayListOfDoubles;
+import vsdk.toolkit.common.Vertex2D;
+import vsdk.toolkit.media.RGBPixel;
+import vsdk.toolkit.media.Image;
+import vsdk.toolkit.environment.geometry.Polygon2D;
+import vsdk.toolkit.environment.geometry._Polygon2DContour;
+
+public class Rasterizer2D extends RenderingElement
+{
+    /**
+    This algorithm implements the Bresenham line algoritm with NO CLIPPING!
+    See [BRES1965].
+    Note that this is a currently naive implementation that makes use
+    of double floating point arithmetic, while original Bresenham algorithm
+    make use of most efficient integer line arithmetic.
+    */
+    public static void drawLine(Image img, int x0, int y0, int x1, int y1, RGBPixel p)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 31. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.